### ISSUE TITLE
networkd-test.py: DnsmasqClientTest.test_resolved_domain_restricted_dns is flaky

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#18357]
Version: 247.2
OS: Debian

This issue is related to #18196.
After disabling `test_resolved_etc_hosts` in Debian, the networkd-test.py autopkgtest now fails in `test_resolved_domain_restricted_dns`.
Recent examples on amd64 are e.g.

https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9898798/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9932260/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9911184/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9907510/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9904176/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9902506/log.gz
https://ci.debian.net/data/autopkgtest/testing/amd64/s/systemd/9900781/log.gz

```
======================================================================
FAIL: test_resolved_domain_restricted_dns (__main__.DnsmasqClientTest)
resolved: domain-restricted DNS servers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest-lxc.xj_j790c/downtmp/build.fFr/src/test/networkd-test.py", line 660, in test_resolved_domain_restricted_dns
    self.assertIn(b'megasearch.net: 192.168.42.1', out)
AssertionError: b'megasearch.net: 192.168.42.1' not found in b'megasearch.net: 207.148.248.143                             -- link: eth0\n\n-- Information acquired via protocol DNS in 3.0ms.\n-- Data is authenticated: no\n'

----------------------------------------------------------------------
Ran 34 tests in 235.361s

FAILED (failures=1, skipped=2)
autopkgtest [21:23:33]: test networkd-test.py: -----------------------]
networkd-test.py     FAIL non-zero exit status 1
```
@martinpitt  any idea?

**Comment by martinpitt** [comment]
@mbiebl: Whoops! `megasearch.net` was meant to be a non-existing bogus domain, and had been for a long time. But it seems some domain grabber recently registered it, and it's an actual thing now:
```
❱❱❱ host megasearch.net
megasearch.net has address 207.148.248.143
```

So at least that test needs to be changed to do `s/megasearch.net/search.example.com/` (I should have used example.com when I wrote it, silly me). It also shows that this test is apparently running with internet connectivity, would probably be better to shield it off. But `*.example.com` is guaranteed to not exist, so that should do fine.

**Comment by mbiebl** [comment]
interestingly, the test seems to succeed sometimes. Is there a race somewhere?

**Comment by mbiebl** [comment]
> So at least that test needs to be changed to do `s/megasearch.net/search.example.com/`

Hmm, if you want a bogus, non-existing domain, this is what I get:

```
$ host example.com
example.com has address 93.184.216.34
example.com has IPv6 address 2606:2800:220:1:248:1893:25c8:1946
example.com mail is handled by 0 .
```

**Comment by martinpitt** [comment]
Ah right, sorry -- see https://www.iana.org/domains/reserved , these are resolvable.

**Comment by mbiebl** [comment]
I'm coming back to this issue. It appears `test_resolved_domain_restricted_dns` is not flaky anymore for me, but completely broken. I locally ran the test 10 times and it failed 10 times.

So I changed the code like you suggested:
```
--- a/test/networkd-test.py
+++ b/test/networkd-test.py
@@ -656,8 +656,8 @@ Domains= ~company ~lab''')
         self.assertIn(b'kettle.cantina.company: 10.241.4.4', out)
 
         # test general domains
-        out = subprocess.check_output(['resolvectl', 'query', 'megasearch.net'])
-        self.assertIn(b'megasearch.net: 192.168.42.1', out)
+        out = subprocess.check_output(['resolvectl', 'query', 'search.example.com'])
+        self.assertIn(b'search.example.com: 192.168.42.1', out)
 
         with open(self.dnsmasq_log) as f:
             general_log = f.read()
@@ -671,8 +671,8 @@ Domains= ~company ~lab''')
         self.assertNotIn('.company', general_log)
 
         # general domains should not be sent to the VPN DNS
-        self.assertRegex(general_log, 'query.*megasearch.net')
-        self.assertNotIn('megasearch.net', vpn_log)
+        self.assertRegex(general_log, 'query.*search.example.com')
+        self.assertNotIn('search.example.com', vpn_log)
 
 #    def test_resolved_etc_hosts(self):
 #        '''resolved queries to /etc/hosts'''
```

But this then fails with 
```
======================================================================
ERROR: test_resolved_domain_restricted_dns (__main__.DnsmasqClientTest)
resolved: domain-restricted DNS servers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest-lxc.g6ymmynw/downtmp/build.icr/real-tree/test/networkd-test.py", line 660, in test_resolved_domain_restricted_dns
    out = subprocess.check_output(['resolvectl', 'query', 'search.example.com'])
  File "/usr/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['resolvectl', 'query', 'search.example.com']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 1 test in 3.733s

FAILED (errors=1)
```

I admit, I don't quite understand what this test is supposed to check. Why should the non-existing domainname `search.example.com` resolve to `192.168.42.1`

**Comment by mbiebl** [comment]
ah, one issue that might contribute to the fact that now test_resolved_domain_restricted_dns fails 100% of the time is that I'm using dnsmasq 2.86. And indeed, with 2.85 it works better.
So I guess this will be something that hits CI as well once updated to a newer dnsmasq.

**Comment by mbiebl** [comment]
Regarding the dnsmasq 2.86 related breakage: A git bisect turned up
https://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=commit;h=12a9aa7c628e2d7dcd34949603848a3fb53fce9c

```
Major rewrite of the DNS server and domain handling code.

This should be largely transparent, but it drastically
improves performance and reduces memory foot-print when
configuring large numbers domains of the form
local=/adserver.com/
or
local=/adserver.com/#

Lookup times now grow as log-to-base-2 of the number of domains,
rather than greater than linearly, as before.
The change makes multiple addresses associated with a domain work
address=/example.com/1.2.3.4
address=/example.com/5.6.7.8
It also handles multiple upstream servers for a domain better; using
the same try/retry alogrithms as non domain-specific servers. This
also applies to DNSSEC-generated queries.

Finally, some of the oldest and gnarliest code in dnsmasq has had
a significant clean-up. It's far from perfect, but it _is_ better.
```
as the first faulty commit.

It appears though, that dnsmasq upstream has already fixed that again in
https://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=commit;h=26bbf5a314d833beaf0f147d24409969f05f3dba
```
Fix --address=/#/...... which was lost in 2.86

A victim of the domain-search rewrite. Apologies.
```
This fix will be in the upcoming 2.87 release.

So, just a heads-up that one should avoid using dnsmasq 2.86
